<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GWF workflow â€“ Title of your project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./notebooks/1_plot_selection.html" rel="next">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./workflow.html"><span class="chapter-title">GWF workflow</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Title of your project</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/munch-group/analysis-template" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Quantifying positive selection on the human X chromosome and its impact on autism</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow.html" class="sidebar-item-text sidebar-link active"><span class="chapter-title">GWF workflow</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Notebooks</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./notebooks/1_plot_selection.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">1_plot_selection.html</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./notebooks/2_runs_test.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">2_runs_test.html</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./notebooks/3_fisher_method.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">3_fisher_method.html</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./notebooks/4_geneinfo.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">4_geneinfo.html</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./notebooks/5_fisher_exact_test.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">5_fisher_exact_test.html</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./notebooks/new_brain_list.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">new_brain_list.html</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-title">GWF workflow</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="af9fa493" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> gwf <span class="im">import</span> Workflow</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys, os, re</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> gwf <span class="im">import</span> Workflow, AnonymousTarget</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> gwf.workflow <span class="im">import</span> collect</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>gwf <span class="op">=</span> Workflow(defaults<span class="op">=</span>{<span class="st">'account'</span>: <span class="st">'ari-intern'</span>})</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># directories</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>out_dir <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps'</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>data_big <span class="op">=</span> <span class="st">'/faststorage/project/ari-intern/data'</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>script_dir <span class="op">=</span> <span class="st">'/faststorage/project/ari-intern/people/ari/ariadna-intern/scripts'</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">'/faststorage/project/ari-intern/people/ari/ariadna-intern/steps/1000genome'</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co"># function that modifies file path</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> modify_path(p, parent<span class="op">=</span><span class="va">None</span>, base<span class="op">=</span><span class="va">None</span>, suffix<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    par, name <span class="op">=</span> os.path.split(p)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    name_no_suffix, suf <span class="op">=</span> os.path.splitext(name)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">type</span>(suffix) <span class="kw">is</span> <span class="bu">str</span>:</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>        suf <span class="op">=</span> suffix</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> parent <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>:</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>        par <span class="op">=</span> parent</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> base <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>:</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>        name_no_suffix <span class="op">=</span> base</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>    new_path <span class="op">=</span> os.path.join(par, name_no_suffix <span class="op">+</span> suf)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">type</span>(suffix) <span class="kw">is</span> <span class="bu">tuple</span>:</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">assert</span> <span class="bu">len</span>(suffix) <span class="op">==</span> <span class="dv">2</span></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>        new_path, nsubs <span class="op">=</span> re.subn(<span class="vs">r'</span><span class="sc">{}</span><span class="vs">$'</span>.<span class="bu">format</span>(suffix[<span class="dv">0</span>]), suffix[<span class="dv">1</span>], new_path)</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">assert</span> nsubs <span class="op">==</span> <span class="dv">1</span>, nsubs</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> new_path</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;function to combine 2 target outputs as an input</span></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> combine(<span class="op">*</span>args, only<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> <span class="bu">all</span>(<span class="bu">len</span>(args[<span class="dv">0</span>]) <span class="op">==</span> <span class="bu">len</span>(args[i]) <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(args)))</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>    combined <span class="op">=</span> []</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(args[<span class="dv">0</span>])):</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>        output_group <span class="op">=</span> {}</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(args)):</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> only:</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>                output_group.update({k: v <span class="cf">for</span> k, v <span class="kw">in</span> args[j].items() <span class="cf">if</span> k <span class="kw">in</span> only})</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>                output_group.update(args[i][j])</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>        combined.append(output_group)</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> combined</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;map of recombination rate across the X chromosome made by DECODE genetics</span></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> decode_genetic_maps(decode_hg38_sexavg_per_gen, genetic_map_chrX):</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> [decode_hg38_sexavg_per_gen]</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> [genetic_map_chrX]</span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'1g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:10:00'</span>}</span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a><span class="ss">    cat </span><span class="sc">{</span>decode_hg38_sexavg_per_gen<span class="sc">}</span><span class="ss"> | tail -n +2 | grep chrX | cut -f 2,4,5 | (echo pos COMBINED_rate Genetic_Map ; cat - ; ) &gt; </span><span class="sc">{</span>genetic_map_chrX<span class="sc">}</span></span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a>decode_hg38_sexavg_per_gen<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>data_big<span class="sc">}</span><span class="ss">/decode_hg38_sexavg_per_gen.tsv'</span></span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a>genetic_map_chrX<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/genetic_map_chrX.tsv'</span></span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a>gwf.target_from_template(<span class="ss">f'decode_genetic_maps'</span>,</span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a>    decode_genetic_maps(decode_hg38_sexavg_per_gen, genetic_map_chrX))</span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a><span class="co"># turn diploid females (XX) into two individual haplotypes (haploid individuals) like males</span></span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> female_haploid(haploid_vcf, chrX_filtered_eagle2_phased, phased_haplotypes):</span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> [haploid_vcf, chrX_filtered_eagle2_phased]</span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> [phased_haplotypes]</span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'01:20:00'</span>}</span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a><span class="ss">    python </span><span class="sc">{</span>haploid_vcf<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>chrX_filtered_eagle2_phased<span class="sc">}</span><span class="ss"> | gzip &gt; </span><span class="sc">{</span>phased_haplotypes<span class="sc">}</span></span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a>haploid_vcf<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>script_dir<span class="sc">}</span><span class="ss">/haploid_vcf.py'</span></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a>chrX_filtered_eagle2_phased<span class="op">=</span><span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000genome/CCDG_14151_B01_GRM_WGS_2020-08-05_chrX.filtered.eagle2-phased.v2.vcf.gz'</span></span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>phased_haplotypes<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes.vcf.gz'</span></span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a>gwf.target_from_template(<span class="ss">f'female_haploid'</span>,</span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a>    female_haploid(haploid_vcf, chrX_filtered_eagle2_phased, phased_haplotypes))</span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a><span class="co"># construct files with haplotype IDs</span></span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> haplotype_id(phased_haplotypes, phased_haplotypes_id):</span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> [phased_haplotypes]</span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> [phased_haplotypes_id]</span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a><span class="ss">    # conda install -c bioconda bcftools</span></span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a><span class="ss">    # conda install openssl   ## to install libcrypto.so.1.0.0 library</span></span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a><span class="ss">    bcftools query -l </span><span class="sc">{</span>phased_haplotypes<span class="sc">}</span><span class="ss"> &gt; </span><span class="sc">{</span>phased_haplotypes_id<span class="sc">}</span></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a><span class="ss">    sleep 5</span></span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a>phased_haplotypes<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes.vcf.gz'</span></span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a>phased_haplotypes_id<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_ids.txt'</span></span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a>gwf.target_from_template(<span class="ss">f'haplotype_id'</span>, haplotype_id(phased_haplotypes, phased_haplotypes_id))</span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;construct populations labels mapping each haplotype to a population</span></span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;(group haplotypes according to the population to which the individuals carrying those haplotypes belong)</span></span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> pop_labels(make_poplabels, phased_haplotypes_id, high_coverage_seq_index, related_high_coverage_seq_index, phased_haplotypes_poplabels):</span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> [make_poplabels, phased_haplotypes_id, high_coverage_seq_index, related_high_coverage_seq_index]</span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> [phased_haplotypes_poplabels]</span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a><span class="ss">    python </span><span class="sc">{</span>make_poplabels<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>phased_haplotypes_id<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>high_coverage_seq_index<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>related_high_coverage_seq_index<span class="sc">}</span><span class="ss"> &gt; </span><span class="sc">{</span>phased_haplotypes_poplabels<span class="sc">}</span><span class="ss"> </span></span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a>make_poplabels<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>script_dir<span class="sc">}</span><span class="ss">/make_poplabels.py'</span></span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a>phased_haplotypes_id<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_ids.txt'</span></span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a>high_coverage_seq_index<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/seq_index/1000G_2504_high_coverage.sequence.index'</span></span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a>related_high_coverage_seq_index<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/seq_index/1000G_698_related_high_coverage.sequence.index'</span></span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a>phased_haplotypes_poplabels<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_poplabels.txt'</span></span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a>gwf.target_from_template(<span class="ss">f'pop_labels'</span>,</span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a>    pop_labels(make_poplabels, phased_haplotypes_id, high_coverage_seq_index, related_high_coverage_seq_index, phased_haplotypes_poplabels))</span>
<span id="cb1-126"><a href="#cb1-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-127"><a href="#cb1-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-128"><a href="#cb1-128" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the function to convert VCF to haps/sample format</span></span>
<span id="cb1-129"><a href="#cb1-129" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> convert_vcf(RelateFileFormats, phased_haplotypes_haps, phased_haplotypes_sample, phased_haplotypes, phased_haplotypes_poplabels):</span>
<span id="cb1-130"><a href="#cb1-130" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> [RelateFileFormats, phased_haplotypes_poplabels, phased_haplotypes]</span>
<span id="cb1-131"><a href="#cb1-131" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> [phased_haplotypes_haps, phased_haplotypes_sample]</span>
<span id="cb1-132"><a href="#cb1-132" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'01:00:00'</span>}</span>
<span id="cb1-133"><a href="#cb1-133" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-134"><a href="#cb1-134" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>RelateFileFormats<span class="sc">}</span><span class="ss"> --mode ConvertFromVcf --haps </span><span class="sc">{</span>phased_haplotypes_haps<span class="sc">}</span><span class="ss"> --sample </span><span class="sc">{</span>phased_haplotypes_sample<span class="sc">}</span><span class="ss"> -i </span><span class="sc">{</span>phased_haplotypes<span class="sc">.</span>replace(<span class="st">'vcf.gz'</span>, <span class="st">''</span>)<span class="sc">}</span><span class="ss"> --poplabels </span><span class="sc">{</span>phased_haplotypes_poplabels<span class="sc">}</span></span>
<span id="cb1-135"><a href="#cb1-135" aria-hidden="true" tabindex="-1"></a><span class="ss">    sleep 20</span></span>
<span id="cb1-136"><a href="#cb1-136" aria-hidden="true" tabindex="-1"></a><span class="ss">    touch </span><span class="sc">{</span>phased_haplotypes_haps<span class="sc">}</span></span>
<span id="cb1-137"><a href="#cb1-137" aria-hidden="true" tabindex="-1"></a><span class="ss">    touch </span><span class="sc">{</span>phased_haplotypes_sample<span class="sc">}</span></span>
<span id="cb1-138"><a href="#cb1-138" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-139"><a href="#cb1-139" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Returning outputs as well</span></span>
<span id="cb1-140"><a href="#cb1-140" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-141"><a href="#cb1-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-142"><a href="#cb1-142" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the file paths for input and output</span></span>
<span id="cb1-143"><a href="#cb1-143" aria-hidden="true" tabindex="-1"></a>RelateFileFormats <span class="op">=</span> <span class="st">'/faststorage/project/ari-intern/people/ari/relate/bin/RelateFileFormats'</span></span>
<span id="cb1-144"><a href="#cb1-144" aria-hidden="true" tabindex="-1"></a>phased_haplotypes_poplabels <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000g_phased_haplotypes_poplabels.txt'</span></span>
<span id="cb1-145"><a href="#cb1-145" aria-hidden="true" tabindex="-1"></a>phased_haplotypes <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000g_phased_haplotypes.vcf.gz'</span></span>
<span id="cb1-146"><a href="#cb1-146" aria-hidden="true" tabindex="-1"></a>phased_haplotypes_haps <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000g_phased_haplotypes.haps'</span></span>
<span id="cb1-147"><a href="#cb1-147" aria-hidden="true" tabindex="-1"></a>phased_haplotypes_sample <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000g_phased_haplotypes.sample'</span></span>
<span id="cb1-148"><a href="#cb1-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-149"><a href="#cb1-149" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating the target using the function</span></span>
<span id="cb1-150"><a href="#cb1-150" aria-hidden="true" tabindex="-1"></a>convert_vcf_target <span class="op">=</span> convert_vcf(RelateFileFormats, phased_haplotypes_haps, phased_haplotypes_sample, phased_haplotypes, phased_haplotypes_poplabels)</span>
<span id="cb1-151"><a href="#cb1-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-152"><a href="#cb1-152" aria-hidden="true" tabindex="-1"></a><span class="co"># Adding the target to the workflow</span></span>
<span id="cb1-153"><a href="#cb1-153" aria-hidden="true" tabindex="-1"></a>gwf.target_from_template(<span class="ss">f'convert_vcf'</span>, convert_vcf_target)</span>
<span id="cb1-154"><a href="#cb1-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-155"><a href="#cb1-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-156"><a href="#cb1-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-157"><a href="#cb1-157" aria-hidden="true" tabindex="-1"></a><span class="co">## start with specific population ##</span></span>
<span id="cb1-158"><a href="#cb1-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-159"><a href="#cb1-159" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;exclude related individuals to avoid biases arising from shared genetic material</span></span>
<span id="cb1-160"><a href="#cb1-160" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> exclude_related(path, population):</span>
<span id="cb1-161"><a href="#cb1-161" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/excluded'</span></span>
<span id="cb1-162"><a href="#cb1-162" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> modify_path(path, parent<span class="op">=</span>output_dir, suffix<span class="op">=</span><span class="st">'_related.txt'</span>)</span>
<span id="cb1-163"><a href="#cb1-163" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'path'</span> : path}</span>
<span id="cb1-164"><a href="#cb1-164" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'path'</span> : output_path}</span>
<span id="cb1-165"><a href="#cb1-165" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-166"><a href="#cb1-166" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-167"><a href="#cb1-167" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-168"><a href="#cb1-168" aria-hidden="true" tabindex="-1"></a><span class="ss">    grep -v '#' </span><span class="sc">{</span>path<span class="sc">}</span><span class="ss"> | cut -f 10 &gt; </span><span class="sc">{</span>output_path<span class="sc">}</span></span>
<span id="cb1-169"><a href="#cb1-169" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-170"><a href="#cb1-170" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-171"><a href="#cb1-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-172"><a href="#cb1-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-173"><a href="#cb1-173" aria-hidden="true" tabindex="-1"></a><span class="co"># find IDs of haplotypes from all other populations so we can exclude them</span></span>
<span id="cb1-174"><a href="#cb1-174" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> ids_other_ppl(path, population):</span>
<span id="cb1-175"><a href="#cb1-175" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/excluded'</span></span>
<span id="cb1-176"><a href="#cb1-176" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> modify_path(path, parent<span class="op">=</span>output_dir, suffix<span class="op">=</span><span class="st">'_non_ppl.txt'</span>)</span>
<span id="cb1-177"><a href="#cb1-177" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'path'</span> : path}</span>
<span id="cb1-178"><a href="#cb1-178" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'path'</span> : output_path}</span>
<span id="cb1-179"><a href="#cb1-179" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-180"><a href="#cb1-180" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-181"><a href="#cb1-181" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-182"><a href="#cb1-182" aria-hidden="true" tabindex="-1"></a><span class="ss">    grep -v </span><span class="sc">{</span>population<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>path<span class="sc">}</span><span class="ss"> | cut -f 1 -d ' ' &gt; </span><span class="sc">{</span>output_path<span class="sc">}</span></span>
<span id="cb1-183"><a href="#cb1-183" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-184"><a href="#cb1-184" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-185"><a href="#cb1-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-186"><a href="#cb1-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-187"><a href="#cb1-187" aria-hidden="true" tabindex="-1"></a><span class="co"># combine excluded files: both related and non population individuals</span></span>
<span id="cb1-188"><a href="#cb1-188" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> combine_files(path, related<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-189"><a href="#cb1-189" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> modify_path(path, base<span class="op">=</span><span class="st">''</span>, suffix<span class="op">=</span><span class="st">'excluded_combined.txt'</span>)</span>
<span id="cb1-190"><a href="#cb1-190" aria-hidden="true" tabindex="-1"></a>    out_dir <span class="op">=</span> modify_path(output_path, base<span class="op">=</span><span class="st">''</span>, suffix<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb1-191"><a href="#cb1-191" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'path'</span>: path, <span class="st">'related'</span>: related}</span>
<span id="cb1-192"><a href="#cb1-192" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'path'</span>: output_path}</span>
<span id="cb1-193"><a href="#cb1-193" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-194"><a href="#cb1-194" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-195"><a href="#cb1-195" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>out_dir<span class="sc">}</span></span>
<span id="cb1-196"><a href="#cb1-196" aria-hidden="true" tabindex="-1"></a><span class="ss">    cat </span><span class="sc">{</span>path<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>related<span class="sc">}</span><span class="ss"> | sort | uniq &gt; </span><span class="sc">{</span>output_path<span class="sc">}</span></span>
<span id="cb1-197"><a href="#cb1-197" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-198"><a href="#cb1-198" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-199"><a href="#cb1-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-200"><a href="#cb1-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-201"><a href="#cb1-201" aria-hidden="true" tabindex="-1"></a><span class="co"># construct a list of excluded individuals</span></span>
<span id="cb1-202"><a href="#cb1-202" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> excluded_list(path, haplotype_id<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-203"><a href="#cb1-203" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> modify_path(path, base<span class="op">=</span><span class="st">''</span>, suffix<span class="op">=</span><span class="st">'excluded_list.txt'</span>)</span>
<span id="cb1-204"><a href="#cb1-204" aria-hidden="true" tabindex="-1"></a>    out_dir <span class="op">=</span> modify_path(output_path, base<span class="op">=</span><span class="st">''</span>, suffix<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb1-205"><a href="#cb1-205" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'path'</span>: path, <span class="st">'haplotype_id'</span>: haplotype_id}</span>
<span id="cb1-206"><a href="#cb1-206" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'exclude_list'</span>: output_path}</span>
<span id="cb1-207"><a href="#cb1-207" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-208"><a href="#cb1-208" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-209"><a href="#cb1-209" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>out_dir<span class="sc">}</span></span>
<span id="cb1-210"><a href="#cb1-210" aria-hidden="true" tabindex="-1"></a><span class="ss">    grep -f </span><span class="sc">{</span>path<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>haplotype_id<span class="sc">}</span><span class="ss"> &gt; </span><span class="sc">{</span>output_path<span class="sc">}</span></span>
<span id="cb1-211"><a href="#cb1-211" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-212"><a href="#cb1-212" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-213"><a href="#cb1-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-214"><a href="#cb1-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-215"><a href="#cb1-215" aria-hidden="true" tabindex="-1"></a><span class="co"># construct a list of only individuals from the population of interest</span></span>
<span id="cb1-216"><a href="#cb1-216" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> pop_labels(exclude_list, poplabels<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-217"><a href="#cb1-217" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/included'</span></span>
<span id="cb1-218"><a href="#cb1-218" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> os.path.join(output_dir, <span class="st">'included_pop_labels.txt'</span>)</span>
<span id="cb1-219"><a href="#cb1-219" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'exclude_list'</span>: exclude_list, <span class="st">'poplabels'</span>: poplabels}</span>
<span id="cb1-220"><a href="#cb1-220" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'pop_label_list'</span>: output_path}</span>
<span id="cb1-221"><a href="#cb1-221" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'10g'</span>, <span class="st">'walltime'</span>: <span class="st">'00:60:00'</span>}</span>
<span id="cb1-222"><a href="#cb1-222" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-223"><a href="#cb1-223" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-224"><a href="#cb1-224" aria-hidden="true" tabindex="-1"></a><span class="ss">    grep -v -f </span><span class="sc">{</span>exclude_list<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>poplabels<span class="sc">}</span><span class="ss"> &gt; </span><span class="sc">{</span>output_path<span class="sc">}</span></span>
<span id="cb1-225"><a href="#cb1-225" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-226"><a href="#cb1-226" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-227"><a href="#cb1-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-228"><a href="#cb1-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-229"><a href="#cb1-229" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;prepare input files for RELATE</span></span>
<span id="cb1-230"><a href="#cb1-230" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> prepare_files(exclude_list, haps<span class="op">=</span><span class="va">None</span>, sample<span class="op">=</span><span class="va">None</span>, ancestor<span class="op">=</span><span class="va">None</span>, mask<span class="op">=</span><span class="va">None</span>, poplabels<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-231"><a href="#cb1-231" aria-hidden="true" tabindex="-1"></a>    directory <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps'</span></span>
<span id="cb1-232"><a href="#cb1-232" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>directory<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/relate'</span></span>
<span id="cb1-233"><a href="#cb1-233" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'haps'</span>: haps, <span class="st">'sample'</span>: sample, <span class="st">'ancestor'</span>: ancestor, <span class="st">'mask'</span>:mask, <span class="st">'poplabels'</span>:poplabels, <span class="st">'exclude_list'</span>:exclude_list}</span>
<span id="cb1-234"><a href="#cb1-234" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> os.path.join(output_dir, <span class="st">'1000g_ppl_phased_haplotypes'</span>)</span>
<span id="cb1-235"><a href="#cb1-235" aria-hidden="true" tabindex="-1"></a>    <span class="co"># outputs: .haps, .sample, .dist (if --mask specified), .poplabels (if remove_ids &amp; poplabels specified), .annot (if poplabels specified)</span></span>
<span id="cb1-236"><a href="#cb1-236" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'haps'</span>: output_path <span class="op">+</span> <span class="st">'.haps'</span>, <span class="st">'sample'</span>: output_path <span class="op">+</span> <span class="st">'.sample'</span>, <span class="st">'dist'</span>: output_path <span class="op">+</span> <span class="st">'.dist'</span>, <span class="st">'poplabels'</span>: output_path <span class="op">+</span> <span class="st">'.poplabels'</span>, <span class="st">'annot'</span>: output_path <span class="op">+</span> <span class="st">'.annot'</span>} </span>
<span id="cb1-237"><a href="#cb1-237" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'20g'</span>, <span class="st">'walltime'</span>: <span class="st">'10:00:00'</span>}</span>
<span id="cb1-238"><a href="#cb1-238" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-239"><a href="#cb1-239" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-240"><a href="#cb1-240" aria-hidden="true" tabindex="-1"></a><span class="ss">    /home/ari/ari-intern/people/ari/relate/scripts/PrepareInputFiles/PrepareInputFiles.sh --haps </span><span class="sc">{</span>haps<span class="sc">}</span><span class="ss"> --sample </span><span class="sc">{</span>sample<span class="sc">}</span><span class="ss"> --ancestor </span><span class="sc">{</span>ancestor<span class="sc">}</span><span class="ss"> --mask </span><span class="sc">{</span>mask<span class="sc">}</span><span class="ss"> --remove_ids </span><span class="sc">{</span>exclude_list<span class="sc">}</span><span class="ss"> --poplabels </span><span class="sc">{</span>poplabels<span class="sc">}</span><span class="ss"> -o </span><span class="sc">{</span>output_path<span class="sc">}</span></span>
<span id="cb1-241"><a href="#cb1-241" aria-hidden="true" tabindex="-1"></a><span class="ss">    sleep 20</span></span>
<span id="cb1-242"><a href="#cb1-242" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-243"><a href="#cb1-243" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-244"><a href="#cb1-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-245"><a href="#cb1-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-246"><a href="#cb1-246" aria-hidden="true" tabindex="-1"></a><span class="co"># compute sfs to make sure singletons are not missing (sanity check)</span></span>
<span id="cb1-247"><a href="#cb1-247" aria-hidden="true" tabindex="-1"></a><span class="co"># zcat 1000g_LWK_phased_haplotypes.haps.gz | cut -d ' ' -f 4- | tr -d -c '1\n' | awk '{ print length; }' | sort -n | uniq -c</span></span>
<span id="cb1-248"><a href="#cb1-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-249"><a href="#cb1-249" aria-hidden="true" tabindex="-1"></a><span class="co"># run the inference of tree sequences using RELATE</span></span>
<span id="cb1-250"><a href="#cb1-250" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> relate(genetic_map, sample_relate<span class="op">=</span><span class="va">None</span>, haps_relate<span class="op">=</span><span class="va">None</span>, annot_relate<span class="op">=</span><span class="va">None</span>, dist_relate<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-251"><a href="#cb1-251" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'/home/ari/ari-intern/people/ari/ariadna-intern/steps/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/relate/run_relate'</span></span>
<span id="cb1-252"><a href="#cb1-252" aria-hidden="true" tabindex="-1"></a>    file_name <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes'</span></span>
<span id="cb1-253"><a href="#cb1-253" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> os.path.join(output_dir, file_name)</span>
<span id="cb1-254"><a href="#cb1-254" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'sample_relate'</span>: sample_relate, <span class="st">'haps_relate'</span>: haps_relate, <span class="st">'annot_relate'</span>: annot_relate, <span class="st">'dist_relate'</span>: dist_relate}</span>
<span id="cb1-255"><a href="#cb1-255" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'anc'</span>: output_path <span class="op">+</span> <span class="st">'.anc'</span>, <span class="st">'mut'</span>: output_path <span class="op">+</span> <span class="st">'.mut'</span>}</span>
<span id="cb1-256"><a href="#cb1-256" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'24g'</span>, <span class="st">'walltime'</span>: <span class="st">'10:00:00'</span>}</span>
<span id="cb1-257"><a href="#cb1-257" aria-hidden="true" tabindex="-1"></a>    <span class="co"># program creates a temporary folder for temporary files and if it already exists relate won't run</span></span>
<span id="cb1-258"><a href="#cb1-258" aria-hidden="true" tabindex="-1"></a>    spec<span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-259"><a href="#cb1-259" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-260"><a href="#cb1-260" aria-hidden="true" tabindex="-1"></a><span class="ss">    cd </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-261"><a href="#cb1-261" aria-hidden="true" tabindex="-1"></a><span class="ss">    rm -rf </span><span class="sc">{</span>file_name<span class="sc">}</span></span>
<span id="cb1-262"><a href="#cb1-262" aria-hidden="true" tabindex="-1"></a><span class="ss">    /home/ari/ari-intern/people/ari/relate/bin/Relate --mode All -m 1.25e-8 -N 20000 --sample </span><span class="sc">{</span>sample_relate<span class="sc">}</span><span class="ss"> --haps </span><span class="sc">{</span>haps_relate<span class="sc">}</span><span class="ss"> --map </span><span class="sc">{</span>genetic_map<span class="sc">}</span><span class="ss"> --annot </span><span class="sc">{</span>annot_relate<span class="sc">}</span><span class="ss"> --dist </span><span class="sc">{</span>dist_relate<span class="sc">}</span><span class="ss"> --memory 20 -o </span><span class="sc">{</span>file_name<span class="sc">}</span></span>
<span id="cb1-263"><a href="#cb1-263" aria-hidden="true" tabindex="-1"></a><span class="ss">    sleep 90</span></span>
<span id="cb1-264"><a href="#cb1-264" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-265"><a href="#cb1-265" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-266"><a href="#cb1-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-267"><a href="#cb1-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-268"><a href="#cb1-268" aria-hidden="true" tabindex="-1"></a><span class="co"># estimate historical population size trajectory from initially inferred tree sequences</span></span>
<span id="cb1-269"><a href="#cb1-269" aria-hidden="true" tabindex="-1"></a><span class="co"># setting --threshold 0. This is so that the branch lengths in all trees are updated for the estimated population size history. </span></span>
<span id="cb1-270"><a href="#cb1-270" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> estimate_ppl_size(anc_size<span class="op">=</span><span class="va">None</span>, mut_size<span class="op">=</span><span class="va">None</span>, poplabels_size<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-271"><a href="#cb1-271" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'/home/ari/ari-intern/people/ari/ariadna-intern/steps/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/relate/run_relate'</span></span>
<span id="cb1-272"><a href="#cb1-272" aria-hidden="true" tabindex="-1"></a>    file_name_input <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes'</span></span>
<span id="cb1-273"><a href="#cb1-273" aria-hidden="true" tabindex="-1"></a>    file_name_output <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes_demog'</span></span>
<span id="cb1-274"><a href="#cb1-274" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> os.path.join(output_dir, file_name_output)</span>
<span id="cb1-275"><a href="#cb1-275" aria-hidden="true" tabindex="-1"></a>    <span class="co">#&nbsp;inputs: inferred .anc/.mut files and a .poplabels file</span></span>
<span id="cb1-276"><a href="#cb1-276" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'anc_size'</span>: anc_size, <span class="st">'mut_size'</span>: mut_size, <span class="st">'poplabels_size'</span>: poplabels_size}</span>
<span id="cb1-277"><a href="#cb1-277" aria-hidden="true" tabindex="-1"></a>    <span class="co"># outputs: two versions of coalescence rates/population sizes are outputted</span></span>
<span id="cb1-278"><a href="#cb1-278" aria-hidden="true" tabindex="-1"></a>    <span class="co">## .coal --&gt; contains coalescence rates and cross-coalescence rates, treating all samples as one population</span></span>
<span id="cb1-279"><a href="#cb1-279" aria-hidden="true" tabindex="-1"></a>    <span class="co">## *.pairwise.coal/.bin --&gt; coalescence rate file and corresponding binary file containing coalescence rates between pairs of samples</span></span>
<span id="cb1-280"><a href="#cb1-280" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'coal'</span>: output_path <span class="op">+</span> <span class="st">'.coal'</span>, <span class="st">'pairwise_coal'</span>: output_path <span class="op">+</span> <span class="st">'.pairwise.coal'</span>, <span class="st">'pairwise_bin'</span>: output_path <span class="op">+</span> <span class="st">'.pairwise.bin'</span>}</span>
<span id="cb1-281"><a href="#cb1-281" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'8g'</span>, <span class="st">'walltime'</span>: <span class="st">'08:00:00'</span>}</span>
<span id="cb1-282"><a href="#cb1-282" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-283"><a href="#cb1-283" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-284"><a href="#cb1-284" aria-hidden="true" tabindex="-1"></a><span class="ss">    cd </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-285"><a href="#cb1-285" aria-hidden="true" tabindex="-1"></a><span class="ss">    rm -rf </span><span class="sc">{</span>file_name_output<span class="sc">}</span></span>
<span id="cb1-286"><a href="#cb1-286" aria-hidden="true" tabindex="-1"></a><span class="ss">    /home/ari/ari-intern/people/ari/relate/scripts/EstimatePopulationSize/EstimatePopulationSize.sh -m 1.25e-8 -N 20000 -i </span><span class="sc">{</span>file_name_input<span class="sc">}</span><span class="ss"> --poplabels </span><span class="sc">{</span>poplabels_size<span class="sc">}</span><span class="ss"> -o </span><span class="sc">{</span>file_name_output<span class="sc">}</span><span class="ss"> --threshold 0 --num_iter 5 --years_per_gen 29 --threads 14 --threshhold 0</span></span>
<span id="cb1-287"><a href="#cb1-287" aria-hidden="true" tabindex="-1"></a><span class="ss">    sleep 20</span></span>
<span id="cb1-288"><a href="#cb1-288" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-289"><a href="#cb1-289" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-290"><a href="#cb1-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-291"><a href="#cb1-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-292"><a href="#cb1-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-293"><a href="#cb1-293" aria-hidden="true" tabindex="-1"></a><span class="co"># detect selection using RELATEs builtin statistic</span></span>
<span id="cb1-294"><a href="#cb1-294" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> detect_selection(anc_selection<span class="op">=</span><span class="va">None</span>, mut_selection<span class="op">=</span><span class="va">None</span>, poplabels_selection<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-295"><a href="#cb1-295" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'/home/ari/ari-intern/people/ari/ariadna-intern/steps/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/relate/run_relate'</span></span>
<span id="cb1-296"><a href="#cb1-296" aria-hidden="true" tabindex="-1"></a>    file_name_input <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes_demog'</span></span>
<span id="cb1-297"><a href="#cb1-297" aria-hidden="true" tabindex="-1"></a>    file_name_output <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes_selection'</span></span>
<span id="cb1-298"><a href="#cb1-298" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> os.path.join(output_dir, file_name_output)</span>
<span id="cb1-299"><a href="#cb1-299" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'anc_selection'</span>: anc_selection, <span class="st">'mut_selection'</span>: mut_selection, <span class="st">'poplabels_selection'</span>: poplabels_selection}</span>
<span id="cb1-300"><a href="#cb1-300" aria-hidden="true" tabindex="-1"></a>    <span class="co">#&nbsp;.freq --&gt; Records the frequency of the derived allele at generations genN .. gen1</span></span>
<span id="cb1-301"><a href="#cb1-301" aria-hidden="true" tabindex="-1"></a>    <span class="co">#&nbsp;.lin --&gt; Records the number of lineages in the tree at generations genN .. gen1 as well as the number of lineages when the mutation had frequency 2</span></span>
<span id="cb1-302"><a href="#cb1-302" aria-hidden="true" tabindex="-1"></a>    <span class="co"># .sele --&gt; Records the log10 p-value for selection evidence at generations genN .. gen1 as well as the log10 p-value when the</span></span>
<span id="cb1-303"><a href="#cb1-303" aria-hidden="true" tabindex="-1"></a>    <span class="co"># mutation had frequency 2. Log10 p-value is set to 1 if mutation had frequency &lt;= 1 at a generation. </span></span>
<span id="cb1-304"><a href="#cb1-304" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'freq_selection'</span>: output_path <span class="op">+</span> <span class="st">'.freq'</span>, <span class="st">'lin_selection'</span>: output_path <span class="op">+</span> <span class="st">'.lin'</span>, <span class="st">'sele_selection'</span>: output_path <span class="op">+</span> <span class="st">'.sele'</span>}</span>
<span id="cb1-305"><a href="#cb1-305" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'20g'</span>, <span class="st">'walltime'</span>: <span class="st">'10:00:00'</span>}</span>
<span id="cb1-306"><a href="#cb1-306" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-307"><a href="#cb1-307" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-308"><a href="#cb1-308" aria-hidden="true" tabindex="-1"></a><span class="ss">    cd </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-309"><a href="#cb1-309" aria-hidden="true" tabindex="-1"></a><span class="ss">    rm -rf </span><span class="sc">{</span>file_name_output<span class="sc">}</span></span>
<span id="cb1-310"><a href="#cb1-310" aria-hidden="true" tabindex="-1"></a><span class="ss">    /home/ari/ari-intern/people/ari/relate/scripts/DetectSelection/DetectSelection.sh -i </span><span class="sc">{</span>file_name_input<span class="sc">}</span><span class="ss"> -m 1.25e-8 --poplabels </span><span class="sc">{</span>poplabels_size<span class="sc">}</span><span class="ss"> -o </span><span class="sc">{</span>file_name_output<span class="sc">}</span></span>
<span id="cb1-311"><a href="#cb1-311" aria-hidden="true" tabindex="-1"></a><span class="ss">    sleep 80</span></span>
<span id="cb1-312"><a href="#cb1-312" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-313"><a href="#cb1-313" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-314"><a href="#cb1-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-315"><a href="#cb1-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-316"><a href="#cb1-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-317"><a href="#cb1-317" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to tree sequence file format (tskit)</span></span>
<span id="cb1-318"><a href="#cb1-318" aria-hidden="true" tabindex="-1"></a><span class="co"># this function converts anc/mut files inferred by Relate into the tree sequence file format used by tskit. In the current</span></span>
<span id="cb1-319"><a href="#cb1-319" aria-hidden="true" tabindex="-1"></a><span class="co">#&nbsp;implementation, each tree is stored with new nodes in the tree sequence file format, leading to no compression. In addition,</span></span>
<span id="cb1-320"><a href="#cb1-320" aria-hidden="true" tabindex="-1"></a><span class="co"># information about how long branches persist, and how many mutations map to a branch are lost by this conversion.</span></span>
<span id="cb1-321"><a href="#cb1-321" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> tree_seq(anc_convert<span class="op">=</span><span class="va">None</span>, mut_convert<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb1-322"><a href="#cb1-322" aria-hidden="true" tabindex="-1"></a>    output_dir <span class="op">=</span> <span class="ss">f'/home/ari/ari-intern/people/ari/ariadna-intern/steps/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/relate/run_relate'</span></span>
<span id="cb1-323"><a href="#cb1-323" aria-hidden="true" tabindex="-1"></a>    file_name_input <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes'</span></span>
<span id="cb1-324"><a href="#cb1-324" aria-hidden="true" tabindex="-1"></a>    file_name_output <span class="op">=</span> <span class="st">'1000g_ppl_phased_haplotypes'</span></span>
<span id="cb1-325"><a href="#cb1-325" aria-hidden="true" tabindex="-1"></a>    output_path <span class="op">=</span> os.path.join(output_dir, file_name_output)</span>
<span id="cb1-326"><a href="#cb1-326" aria-hidden="true" tabindex="-1"></a>    <span class="co"># inputs: .anc (genealogical relationships info) and .mut (mutations info)</span></span>
<span id="cb1-327"><a href="#cb1-327" aria-hidden="true" tabindex="-1"></a>    inputs <span class="op">=</span> {<span class="st">'anc_convert'</span>: anc_convert, <span class="st">'mut_convert'</span>: mut_convert}</span>
<span id="cb1-328"><a href="#cb1-328" aria-hidden="true" tabindex="-1"></a>    <span class="co"># outputs: .trees (combination of both inputs)</span></span>
<span id="cb1-329"><a href="#cb1-329" aria-hidden="true" tabindex="-1"></a>    outputs <span class="op">=</span> {<span class="st">'trees_convert'</span>: output_path <span class="op">+</span> <span class="st">'.trees'</span>}</span>
<span id="cb1-330"><a href="#cb1-330" aria-hidden="true" tabindex="-1"></a>    options <span class="op">=</span> {<span class="st">'memory'</span>: <span class="st">'8g'</span>, <span class="st">'walltime'</span>: <span class="st">'04:00:00'</span>}</span>
<span id="cb1-331"><a href="#cb1-331" aria-hidden="true" tabindex="-1"></a>    spec <span class="op">=</span> <span class="ss">f'''</span></span>
<span id="cb1-332"><a href="#cb1-332" aria-hidden="true" tabindex="-1"></a><span class="ss">    mkdir -p </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-333"><a href="#cb1-333" aria-hidden="true" tabindex="-1"></a><span class="ss">    cd </span><span class="sc">{</span>output_dir<span class="sc">}</span></span>
<span id="cb1-334"><a href="#cb1-334" aria-hidden="true" tabindex="-1"></a><span class="ss">    rm -rf </span><span class="sc">{</span>file_name_output<span class="sc">}</span></span>
<span id="cb1-335"><a href="#cb1-335" aria-hidden="true" tabindex="-1"></a><span class="ss">    /home/ari/ari-intern/people/ari/relate/bin/RelateFileFormats --mode ConvertToTreeSequence -i </span><span class="sc">{</span>file_name_input<span class="sc">}</span><span class="ss"> -o </span><span class="sc">{</span>file_name_output<span class="sc">}</span></span>
<span id="cb1-336"><a href="#cb1-336" aria-hidden="true" tabindex="-1"></a><span class="ss">    '''</span></span>
<span id="cb1-337"><a href="#cb1-337" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> AnonymousTarget(inputs<span class="op">=</span>inputs, outputs<span class="op">=</span>outputs, options<span class="op">=</span>options, spec<span class="op">=</span>spec)</span>
<span id="cb1-338"><a href="#cb1-338" aria-hidden="true" tabindex="-1"></a>                    </span>
<span id="cb1-339"><a href="#cb1-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-340"><a href="#cb1-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-341"><a href="#cb1-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-342"><a href="#cb1-342" aria-hidden="true" tabindex="-1"></a><span class="co">#populations = ['GWD', 'ESN', 'MSL', 'YRI', 'LWK'] # african ancestry --&gt; DONE</span></span>
<span id="cb1-343"><a href="#cb1-343" aria-hidden="true" tabindex="-1"></a>populations <span class="op">=</span> [<span class="st">'KHV'</span>]</span>
<span id="cb1-344"><a href="#cb1-344" aria-hidden="true" tabindex="-1"></a><span class="co">#populations = ['GBR', 'FIN', 'IBS',# 'TSI'] + PUR (puerto ricans)#&nbsp;european ancestry --&gt; DONE</span></span>
<span id="cb1-345"><a href="#cb1-345" aria-hidden="true" tabindex="-1"></a><span class="co">#populations = ['CDX', 'CHB', 'CHS', #'JPT', 'KHV'] # east asian ancestry</span></span>
<span id="cb1-346"><a href="#cb1-346" aria-hidden="true" tabindex="-1"></a><span class="co">#populations = ['LWK', 'GWD', 'ESN', 'MSL', 'YRI', 'GBR', 'FIN', 'IBS', 'TSI', 'CDX', 'CHB', 'CHS', 'JPT', 'KHV']</span></span>
<span id="cb1-347"><a href="#cb1-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-348"><a href="#cb1-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-349"><a href="#cb1-349" aria-hidden="true" tabindex="-1"></a><span class="co"># append a unique identifier to each target name to ensure they are distinct (name = ...{population})</span></span>
<span id="cb1-350"><a href="#cb1-350" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> population <span class="kw">in</span> populations:</span>
<span id="cb1-351"><a href="#cb1-351" aria-hidden="true" tabindex="-1"></a>    <span class="co"># exlcude related</span></span>
<span id="cb1-352"><a href="#cb1-352" aria-hidden="true" tabindex="-1"></a>    input_related <span class="op">=</span> [(<span class="ss">f'</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/seq_index/1000G_698_related_high_coverage.sequence.index'</span>, population)]</span>
<span id="cb1-353"><a href="#cb1-353" aria-hidden="true" tabindex="-1"></a>    related_target <span class="op">=</span> gwf.<span class="bu">map</span>(exclude_related, input_related, name<span class="op">=</span><span class="ss">f"exclude_related_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-354"><a href="#cb1-354" aria-hidden="true" tabindex="-1"></a>    related <span class="op">=</span> related_target.outputs[<span class="dv">0</span>]  <span class="co"># list</span></span>
<span id="cb1-355"><a href="#cb1-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-356"><a href="#cb1-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-357"><a href="#cb1-357" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get ids for other populations</span></span>
<span id="cb1-358"><a href="#cb1-358" aria-hidden="true" tabindex="-1"></a>    input_other_ppl <span class="op">=</span> [(<span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_poplabels.txt'</span>, population)]</span>
<span id="cb1-359"><a href="#cb1-359" aria-hidden="true" tabindex="-1"></a>    other_ppl_target <span class="op">=</span> gwf.<span class="bu">map</span>(ids_other_ppl, input_other_ppl, name<span class="op">=</span><span class="ss">f"ids_other_ppl_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-360"><a href="#cb1-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-361"><a href="#cb1-361" aria-hidden="true" tabindex="-1"></a>    <span class="co"># combine related and other populations</span></span>
<span id="cb1-362"><a href="#cb1-362" aria-hidden="true" tabindex="-1"></a>    combine_target <span class="op">=</span> gwf.<span class="bu">map</span>(combine_files, other_ppl_target.outputs, extra <span class="op">=</span> {<span class="st">'related'</span>:related}, name<span class="op">=</span><span class="ss">f"combine_files_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-363"><a href="#cb1-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-364"><a href="#cb1-364" aria-hidden="true" tabindex="-1"></a>    <span class="co">#&nbsp;list of excluded</span></span>
<span id="cb1-365"><a href="#cb1-365" aria-hidden="true" tabindex="-1"></a>    haplotype_ids <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_ids.txt'</span></span>
<span id="cb1-366"><a href="#cb1-366" aria-hidden="true" tabindex="-1"></a>    exclude_list_target <span class="op">=</span> gwf.<span class="bu">map</span>(excluded_list, combine_target.outputs, extra <span class="op">=</span> {<span class="st">'haplotype_id'</span>:haplotype_ids}, name<span class="op">=</span><span class="ss">f"excluded_list_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-367"><a href="#cb1-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-368"><a href="#cb1-368" aria-hidden="true" tabindex="-1"></a>    <span class="co"># list of included</span></span>
<span id="cb1-369"><a href="#cb1-369" aria-hidden="true" tabindex="-1"></a>    poplabels <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_poplabels.txt'</span></span>
<span id="cb1-370"><a href="#cb1-370" aria-hidden="true" tabindex="-1"></a>    pop_labels_target <span class="op">=</span> gwf.<span class="bu">map</span>(pop_labels, exclude_list_target.outputs, extra <span class="op">=</span> {<span class="st">'poplabels'</span>:poplabels}, name<span class="op">=</span><span class="ss">f"pop_labels_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-371"><a href="#cb1-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-372"><a href="#cb1-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-373"><a href="#cb1-373" aria-hidden="true" tabindex="-1"></a>    <span class="co"># RELATE DIRECTORY !</span></span>
<span id="cb1-374"><a href="#cb1-374" aria-hidden="true" tabindex="-1"></a>    relate_dir <span class="op">=</span> <span class="ss">f'/home/ari/ari-intern/people/ari/ariadna-intern/steps/</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">/relate'</span> <span class="co"># relate directory</span></span>
<span id="cb1-375"><a href="#cb1-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-376"><a href="#cb1-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-377"><a href="#cb1-377" aria-hidden="true" tabindex="-1"></a>    <span class="co"># PREPARE INPUT</span></span>
<span id="cb1-378"><a href="#cb1-378" aria-hidden="true" tabindex="-1"></a>    haps <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000g_phased_haplotypes.haps'</span></span>
<span id="cb1-379"><a href="#cb1-379" aria-hidden="true" tabindex="-1"></a>    sample <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/1000g_phased_haplotypes.sample'</span></span>
<span id="cb1-380"><a href="#cb1-380" aria-hidden="true" tabindex="-1"></a>    ancestor <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/homo_sapiens_ancestor_GRCh38/homo_sapiens_ancestor_X.fa'</span></span>
<span id="cb1-381"><a href="#cb1-381" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>data_dir<span class="sc">}</span><span class="ss">/20160622.chrX.mask.fasta'</span></span>
<span id="cb1-382"><a href="#cb1-382" aria-hidden="true" tabindex="-1"></a>    poplabels <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>out_dir<span class="sc">}</span><span class="ss">/1000g_phased_haplotypes_poplabels.txt'</span></span>
<span id="cb1-383"><a href="#cb1-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-384"><a href="#cb1-384" aria-hidden="true" tabindex="-1"></a>    prepare_target <span class="op">=</span> gwf.<span class="bu">map</span>(prepare_files, exclude_list_target.outputs, </span>
<span id="cb1-385"><a href="#cb1-385" aria-hidden="true" tabindex="-1"></a>                            extra <span class="op">=</span> {<span class="st">'haps'</span>: haps, <span class="st">'sample'</span>: sample, <span class="st">'ancestor'</span>: ancestor, <span class="st">'mask'</span>:mask, <span class="st">'poplabels'</span>:poplabels}, name<span class="op">=</span><span class="ss">f"prepare_files_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-386"><a href="#cb1-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-387"><a href="#cb1-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-388"><a href="#cb1-388" aria-hidden="true" tabindex="-1"></a>    <span class="co"># RUN RELATE</span></span>
<span id="cb1-389"><a href="#cb1-389" aria-hidden="true" tabindex="-1"></a>    sample_relate <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/1000g_ppl_phased_haplotypes.sample.gz'</span></span>
<span id="cb1-390"><a href="#cb1-390" aria-hidden="true" tabindex="-1"></a>    haps_relate <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/1000g_ppl_phased_haplotypes.haps.gz'</span></span>
<span id="cb1-391"><a href="#cb1-391" aria-hidden="true" tabindex="-1"></a>    genetic_map <span class="op">=</span> <span class="st">'/home/ari/ari-intern/people/ari/ariadna-intern/steps/genetic_map_chrX.tsv'</span></span>
<span id="cb1-392"><a href="#cb1-392" aria-hidden="true" tabindex="-1"></a>    annot_relate <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/1000g_ppl_phased_haplotypes.annot'</span></span>
<span id="cb1-393"><a href="#cb1-393" aria-hidden="true" tabindex="-1"></a>    dist_relate <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/1000g_ppl_phased_haplotypes.dist.gz'</span></span>
<span id="cb1-394"><a href="#cb1-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-395"><a href="#cb1-395" aria-hidden="true" tabindex="-1"></a>    run_relate_target <span class="op">=</span> gwf.<span class="bu">map</span>(relate, [genetic_map], extra <span class="op">=</span> {<span class="st">'haps_relate'</span>: haps_relate, <span class="st">'sample_relate'</span>: sample_relate, <span class="st">'annot_relate'</span>: annot_relate, <span class="st">'dist_relate'</span>: dist_relate}, name<span class="op">=</span><span class="ss">f"relate_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-396"><a href="#cb1-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-397"><a href="#cb1-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-398"><a href="#cb1-398" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ESTIMATE POPULATION SIZES</span></span>
<span id="cb1-399"><a href="#cb1-399" aria-hidden="true" tabindex="-1"></a>    anc_size <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/run_relate/1000g_ppl_phased_haplotypes.anc'</span></span>
<span id="cb1-400"><a href="#cb1-400" aria-hidden="true" tabindex="-1"></a>    mut_size  <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/run_relate/1000g_ppl_phased_haplotypes.mut'</span></span>
<span id="cb1-401"><a href="#cb1-401" aria-hidden="true" tabindex="-1"></a>    poplabels_size <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/1000g_ppl_phased_haplotypes.poplabels'</span></span>
<span id="cb1-402"><a href="#cb1-402" aria-hidden="true" tabindex="-1"></a>    ppl_size_target <span class="op">=</span> gwf.<span class="bu">map</span>(estimate_ppl_size, [anc_size], extra <span class="op">=</span> {<span class="st">'mut_size'</span>: mut_size, <span class="st">'poplabels_size'</span>: poplabels_size}, name<span class="op">=</span><span class="ss">f"estimate_ppl_size_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-403"><a href="#cb1-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-404"><a href="#cb1-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-405"><a href="#cb1-405" aria-hidden="true" tabindex="-1"></a>    <span class="co"># DETECT SELECTION</span></span>
<span id="cb1-406"><a href="#cb1-406" aria-hidden="true" tabindex="-1"></a>    anc_selection <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/run_relate/1000g_ppl_phased_haplotypes_demog.anc.gz'</span></span>
<span id="cb1-407"><a href="#cb1-407" aria-hidden="true" tabindex="-1"></a>    mut_selection  <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/run_relate/1000g_ppl_phased_haplotypes_demog.mut.gz'</span></span>
<span id="cb1-408"><a href="#cb1-408" aria-hidden="true" tabindex="-1"></a>    poplabels_selection <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/1000g_ppl_phased_haplotypes.poplabels'</span></span>
<span id="cb1-409"><a href="#cb1-409" aria-hidden="true" tabindex="-1"></a>    detect_selection_target <span class="op">=</span> gwf.<span class="bu">map</span>(detect_selection, [anc_selection], extra <span class="op">=</span> {<span class="st">'mut_selection'</span>: mut_selection, <span class="st">'poplabels_selection'</span>: poplabels_selection}, name<span class="op">=</span><span class="ss">f"detect_selection_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb1-410"><a href="#cb1-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-411"><a href="#cb1-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-412"><a href="#cb1-412" aria-hidden="true" tabindex="-1"></a>   <span class="co"># CONVERT TO TREE SEQUENCE FILE FORMAT (tskit)</span></span>
<span id="cb1-413"><a href="#cb1-413" aria-hidden="true" tabindex="-1"></a>    anc_convert <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/run_relate/1000g_ppl_phased_haplotypes.anc'</span></span>
<span id="cb1-414"><a href="#cb1-414" aria-hidden="true" tabindex="-1"></a>    mut_convert  <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>relate_dir<span class="sc">}</span><span class="ss">/run_relate/1000g_ppl_phased_haplotypes.mut'</span></span>
<span id="cb1-415"><a href="#cb1-415" aria-hidden="true" tabindex="-1"></a>    tree_seq_target <span class="op">=</span> gwf.<span class="bu">map</span>(tree_seq, [anc_convert], extra <span class="op">=</span> {<span class="st">'mut_convert'</span>: mut_convert}, name<span class="op">=</span><span class="ss">f"tree_convert_</span><span class="sc">{</span>population<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link" aria-label="Quantifying positive selection on the human X chromosome and its impact on autism">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-title">Quantifying positive selection on the human X chromosome and its impact on autism</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./notebooks/1_plot_selection.html" class="pagination-link" aria-label="1_plot_selection.html">
        <span class="nav-page-text"><span class="chapter-title">1_plot_selection.html</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions"><ul><li><a href="https://github.com/munch-group/analysis-template/edit/main/workflow.py" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/munch-group/analysis-template/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li><li><a href="https://github.com/munch-group/analysis-template/blob/main/workflow.py" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>